---
title: "Generally Applicable Gene-set Enrichment (GAGE)"
author: "Oliver Hölsken"
date: "12.12.2021"
output:
  html_document: default
  pdf_document: default
  github_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The github repository can be found [here](https://github.com/ohoelske/MDS_KW4_GSE)

# Set up Working environment

## Installations
```{r, message=FALSE,warning=FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(version = "3.14")
```

```{r, message=FALSE,warning=FALSE}
#BiocManager::install(c("gage", "org.Hs.eg.db", "clusterProfiler"))
```

```{r, message=FALSE,warning=FALSE}
#install.packages("msigdbr")
#install.packages("openxlsx")
```

## Load libraries
```{r,  message=FALSE,warning=FALSE}
library(BiocManager)
library(ggplot2)
library(ggrepel)
library(dplyr)
library(tibble)
library(readr)
library(tidyr)
library(tidyverse)
library(msigdbr)
library(clusterProfiler)
library(fgsea)
library(gage)
library(org.Hs.eg.db)
```

## Tasks

### Part 1: Getting familiar with Gene-sets

The Hallmark get can be found [here](https://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=H).

1. Download Hallmark gene-sets using the `msigdbr` R package.
```{r}
#Check available gene sets using msigdbr_collections()
msigdbr_collections()
```

```{r}
# get Human H (Hallmark) gene sets
h_gene_set <- msigdbr(species = "human", 
                      category = "H")
head(h_gene_set)
# get Human Reactome gene sets
r_gene_set <- msigdbr(species = "human", 
                      category = "C2",
                      subcategory = "CP:REACTOME")
head(h_gene_set)
```
2.	Look at Hallmark: check the number of gene-sets, number of genes per gene set…

```{r}
# Create a nested data (list-column of a data frame) 
n_h_gene_set <- h_gene_set %>%
  group_by(gs_name) %>%
  nest()
# Count no of rows = gene sets
nrow(n_h_gene_set)
```

```{r}
n_h_gene_set
```



The Hallmark Datasets contains 50 Gene-sets
```{r}
#Extract the Gene-set names 
Gene_set_name <- n_h_gene_set$gs_name
head(Gene_set_name)
```

```{r}
#Define no of Genes per Gene set by using the purrr::map function
Genes_per_Gene_set <- map_int(n_h_gene_set$data, ~nrow(.))
```

```{r}
# Generate df
GS_genes <- data.frame(Gene_set_name, Genes_per_Gene_set)
head(GS_genes)
```

```{r}
knitr::kable(GS_genes,
caption = "Genes per Hallmark Geneset")
```

3.	Prepare a figure (e.g. barplot) showing the number of genes per gene set.
```{r}
#Transform data frame to remove "Hallmark" string
GS_genes$Gene_set_name <- 
  gsub("HALLMARK_","",as.character(GS_genes$Gene_set_name))
head(GS_genes)
```

```{r}
#Change factor lavels to order according to no of genes per Gene set
GS_genes$Gene_set_name <- factor(GS_genes$Gene_set_name, 
                                 levels = GS_genes$Gene_set_name
                                 [order(GS_genes$Genes_per_Gene_set)])
```

```{r fig.height=15, fig.width = 10, fig.align='center', fig.cap='Genes per gene-set'}
#Create the Bar Plot
x <- ggplot(GS_genes, aes(x = Gene_set_name, y = Genes_per_Gene_set)) + 
  theme_bw() + 
  geom_bar(stat = "identity", fill="#E69F00") +
  coord_flip()
x + ggtitle("Genes per gene set of the Hallmark Gene set") +
  xlab("HALLMARK Gene set") +
  ylab("Genes per Gene set")
```
```{r, warning=FALSE}
#Save as PDF
ggsave(
  file = "GenesPerGeneset.pdf",
  plot = last_plot(),
  path = NULL,
  scale = 1,
  width = 10,
  height = 15,
  units = "in",
  dpi = "retina",
  limitsize = TRUE,
  bg = NULL,
)
```


4. Convert the hallmark `data.frame` into a named list (to be used as “gsets” parameter with gage)
(For this, the `readList` function was used from the downloaded `.gmt` files. This is probably not the most elegant way but it served its purpose I hope.)
```{r}
#Read in Database information as List
hallmarks_symbols <- readList("h.all.v7.4.symbols.gmt")
hallmarks_entrez <- readList("h.all.v7.4.entrez.gmt")
```

5. Do the same with the `Reactome` database
```{r}
reactome_entrez <- readList("c2.cp.reactome.v7.4.entrez.gmt")
reactome_symbol <- readList("c2.cp.reactome.v7.4.symbols.gmt")
```


### Part 2: 

Now we are preparing the gene-expression data.

1.	Load the hallmark list you have created
```{r}
hallmark_list <- hallmarks_entrez
```

2.	Load the log2 TPM matrix.
```{r}
expr.data <- read.delim("TCGA_GBM_log2TPM.txt")
expr.data <- as.data.frame(expr.data)
head(expr.data)
```

```{r}
#convert the data.frame into a matrix as to speed up the computing
expr.data.matr <- as.matrix(expr.data)
str(expr.data.matr)
```
3. Check the overlap between the database and the genes within the RNA matrix

```{r}
head(hallmark_list[[1]]); head(rownames(expr.data.matr))
```
gene ID types are the same in the gene sets `hallmark_list` and expression data`expr.data.mat`

4.  Create an annotation `data.frame` from the column names of the RNA matrix (i.e. assign each sample to either primary, recurrent or normal condition).
```{r}
my_sample_col <- data.frame(condition = rep(c("Normal", "Primary", "Recurrent"), 
                                            c(5,5,5)))
row.names(my_sample_col) <- colnames(expr.data)
my_sample_col
```

```{r}
# Check if all rownames from annotated sample column are equal to colnames of the expression data
all(rownames(my_sample_col) == colnames(expr.data))
```

```{r}
head(expr.data)
```


5.	Calculate the average log2 foldchange between primary and normal, recurrent and normal, recurrent and primary.
```{r}
#calculate the mean for each gene per group
normal_mean <- apply(expr.data[,1:5],1, mean)
primary_mean <- apply(expr.data[,6:10],1, mean)
recurrent_mean <- apply(expr.data[,11:15],1, mean)
```

```{r}
#calculate the fold change between the groups
## Normal/Primary
foldchange_nor_primary <- normal_mean - primary_mean 
## Normal/Recurrent
foldchange_nor_recurrent <- normal_mean - recurrent_mean
## Primary/Recurrent
foldchange_pri_recurrent <- primary_mean - recurrent_mean
```

```{r}
hist(foldchange_nor_primary, 
     xlab = "log2 Fold Change (Normal vs Primary)")
```

```{r}
hist(foldchange_nor_recurrent, 
     xlab = "log2 Fold Change (Normal vs Recurrent)")
```

```{r}
hist(foldchange_pri_recurrent, 
     xlab = "log2 Fold Change (Primary vs Recurrent)")
```

```{r}
#Add log2FC for normal/tumor
expr.data.fc <- expr.data
expr.data.fc$log2FC_nor_pri <- foldchange_nor_primary
expr.data.fc$log2FC_nor_rec <- foldchange_nor_recurrent
expr.data.fc$log2FC_pri_recurrent <- foldchange_pri_recurrent
expr.data.fc <- rownames_to_column(expr.data.fc, var = "gene_ID")
```

```{r}
# add Gene_Symbol
library("gage")
data(egSymb)
expr.data.fc$Gene_Symbol <- eg2sym(expr.data.fc$gene_ID) 
head(expr.data.fc)
```


```{r}
#Select genes, which are upregulated in expression matrix (normal vs. primary)
genes.up_pri_no <- subset(expr.data.fc, 
                          log2FC_nor_pri > 0, 
                          select = c("Gene_Symbol", "gene_ID"))
head(genes.up_pri_no)
up_genes <- genes.up_pri_no$gene_ID
```

```{r}
#Select genes, which are downregulated in expxression matrix (normal vs. primary)
genes.down_pri_no <- subset(expr.data.fc, 
                            log2FC_nor_pri < 0,
                            select = c("Gene_Symbol", "gene_ID"))
head(genes.down_pri_no)
down_genes <- genes.down_pri_no$gene_ID
```

```{r}
#Select genes, which are up/downregulated in epxression matrix (normal vs. recurrent)
## Upregulated
genes.up_nor_rec <- subset(expr.data.fc, 
                           log2FC_nor_rec >0, 
                           select = c("Gene_Symbol", "gene_ID"))
## down
genes.down_nor_rec <- subset(expr.data.fc, 
                           log2FC_nor_rec <0, 
                           select = c("Gene_Symbol", "gene_ID"))
#Select genes, which are up/downregulated in epxression matrix (primary vs. recurrent)
## Upregulated
genes.up_pri_rec <- subset(expr.data.fc, 
                           log2FC_pri_recurrent >0, 
                           select = c("Gene_Symbol", "gene_ID"))
## down
genes.down_pri_rec <- subset(expr.data.fc, 
                           log2FC_pri_recurrent <0, 
                           select = c("Gene_Symbol", "gene_ID"))
```


6. Use `Gage` (unpaired test) to identify differentially UP- and DOWN-regulated gene-sets (using `Hallmark` and `Reactome` gene-sets separately). Make the following comparisons: primary vs. normal, recurrent vs. normal, recurrent vs. primary.
```{r}
#Definition of gage inputs
cn <- colnames(expr.data)
normal <- grep('Normal',cn, ignore.case =T)
primary <- grep('Primary',cn, ignore.case =T)
recurrent <- grep('Recurrent',cn, ignore.case =T)
```

```{r}
#Hallmarks
gage_H_pri_no = gage(exprs = expr.data, 
                     gsets = hallmark_list, 
                     set.size=c(10,500), 
                     same.dir = TRUE, 
                     compare ="unpaired", 
                     ref = normal, 
                     samp = primary) 

gage_H_rec_no = gage(exprs = expr.data, 
                     gsets = hallmark_list, 
                     set.size=c(10,500), 
                     same.dir = TRUE, 
                     compare ="unpaired", 
                     ref = normal, 
                     samp = recurrent)

gage_H_pri_rec = gage(exprs = expr.data, 
                      gsets = hallmark_list, 
                      set.size=c(10,500), 
                      same.dir = TRUE, 
                      compare ="unpaired", 
                      ref = primary, 
                      samp = recurrent)
```

```{r}
#Reactome database
gage_R_pri_no = gage(exprs = expr.data, 
                     gsets = reactome_entrez, 
                     set.size=c(10,500), 
                     same.dir = TRUE, 
                     compare ="unpaired", 
                     ref = normal, 
                     samp = primary) 

gage_R_rec_no = gage(exprs = expr.data, 
                     gsets = reactome_entrez, 
                     set.size=c(10,500), 
                     same.dir = TRUE, 
                     compare ="unpaired", 
                     ref = normal, 
                     samp = recurrent)

gage_R_pri_rec = gage(exprs = expr.data, 
                      gsets = reactome_entrez, 
                      set.size=c(10,500), 
                      same.dir = TRUE, 
                      compare ="unpaired", 
                      ref = primary, 
                      samp = recurrent)
```


```{r}
#Determine the significant gene sets by using the sigGeneSetFunction.
no_pri_H.sig<-sigGeneSet(gage_H_pri_no, cutoff = 0.1)
no_rec_H.sig<-sigGeneSet(gage_H_rec_no, cutoff = 0.1)
pri_rec_H.sig<-sigGeneSet(gage_H_pri_rec, cutoff = 0.1)
#Do the same for the reactome dataset
no_pri_R.sig<-sigGeneSet(gage_R_pri_no, cutoff = 0.1)
no_rec_R.sig<-sigGeneSet(gage_R_rec_no, cutoff = 0.1)
pri_rec_R.sig<-sigGeneSet(gage_R_pri_rec, cutoff = 0.1)
```


```{r}
#UpregulatesGene sets Normal vs. Primary as data frame
df_no_pri_H.sig <- as.data.frame(no_pri_H.sig$greater)
df_no_pri_H.sig <- subset(df_no_pri_H.sig,
                          select = c("p.val","q.val","set.size"))
head(df_no_pri_H.sig)
```
```{r}
#DownregulatesGene sets Normal vs. Primary as data frame
df_no_pri_H.sig.down <- as.data.frame(no_pri_H.sig$less)
df_no_pri_H.sig.down <- subset(df_no_pri_H.sig.down,
                          select = c("p.val","q.val","set.size"))
head(df_no_pri_H.sig.down)
```

```{r}
#Create annotated Hallmark list
library(purrr)
#Upregulated genes
intersect_up_genes <- function(x) intersect(x, up_genes)
annotated_hallmarks_up <- map(hallmark_list, intersect_up_genes)
#Downregulated genes
intersect_down_genes <- function(x) intersect(x, down_genes)
annotated_hallmarks_down  <- map(hallmark_list, intersect_down_genes)
```

```{r}
#UP df
up_list <- t(map_dfc(annotated_hallmarks_up, ~length(.)))
up_df <- as.data.frame(up_list)
colnames(up_df)[colnames(up_df ) == "V1"] <- "nb"
up_df 
```

```{r}
library(tibble)
ned_df_2 <- rownames_to_column(df_no_pri_H.sig)
ned_df_3 <- rownames_to_column(up_df)
```

```{r}
#Create the final table with upregulated genes
final_df_up <- inner_join(ned_df_2, ned_df_3, by = "rowname", copy = FALSE)
colnames(final_df_up )[1] <- "Gene_set"
final_df_up$log10p.val <- -log10(final_df_up$p.val)
head(final_df_up)
```
```{r}
knitr::kable(final_df_up,
caption = "Significantly upregulated Gene sets (Normal / Tumor)")
```

Perform for downregulated genes
```{r}
#down df
down_list <- t(map_dfc(annotated_hallmarks_down, ~length(.)))
down_df <- as.data.frame(down_list)
colnames(down_df)[colnames(down_df ) == "V1"] <- "nb"
down_df 
```

```{r}
ned_df_4 <- rownames_to_column(df_no_pri_H.sig.down)
ned_df_5 <- rownames_to_column(down_df)
```

```{r}
#Create the final table with downregulated genes
final_df_down <- inner_join(ned_df_4, ned_df_5, by = "rowname", copy = FALSE)
colnames(final_df_down)[1] <- "Gene_set"
final_df_down$log10p.val <- -log10(final_df_down$p.val)
```

```{r}
knitr::kable(final_df_down,
caption = "Significantly downregulated Gene sets (Normal / Tumor)")
```

Export results
```{r}
#library opelxlsx
library(openxlsx)

#Write xlsx of DEG up- and downregulated (with two sheets)
list_of_datasets <- list("upregulated" = final_df_up, 
                         "downregulated" = final_df_down)

write.xlsx(list_of_datasets, 
           file = "Sig_UP_DOWN_Gene_sets_Tumor_Normal.xlsx", 
           asTable = F, 
           firstRow = T, 
           headerStyle = createStyle(textDecoration = 'bold'), 
           keepNA = F, 
           rowNames = F, 
           overwrite = T)
```




7. Annotate the significantly regulated gene-sets with genes (entrez IDs and symbols). Check org.Hs.eg.db R package.

#### To be continued

8.	Visualise the most significant gene-sets (e.g. using barplots or heatmaps)

```{r}
#Change factor lavels to order according to no of genes per Gene set
final_df_up$Gene_set <- factor(final_df_up$Gene_set, 
                                 levels = final_df_up$Gene_set 
                                 [order(final_df_up$log10p.val)])
```


```{r fig.height=10, fig.width = 10, fig.align='center', fig.cap='Sign. upregulated Gene sets'}
#Create the Bar Plot
p <- ggplot(final_df_up, 
       aes(x = Gene_set, y = log10p.val)) + 
  theme_bw() + 
  geom_bar(stat = "identity", fill = "#FF6666") +
  coord_flip() 

p + ggtitle ("Signficantly upregulated Gene sets (Normal vs Tumor") +
  ylab("-log10pvalue")
```

```{r, warning=FALSE}
#Save as PDF
ggsave(
  file = "Upregulated_genes_bar.pdf",
  plot = last_plot(),
  path = NULL,
  scale = 1,
  width = 10,
  height = 15,
  units = "in",
  dpi = "retina",
  limitsize = TRUE,
  bg = NULL,
)
```


```{r}
#Change factor lavels to order according to no of genes per Gene set
final_df_down$Gene_set <- factor(final_df_down$Gene_set, 
                                 levels = final_df_down$Gene_set 
                                 [order(final_df_down$log10p.val)])
```

```{r fig.height=5, fig.width = 10, fig.align='center', fig.cap='Sign. downregulated Gene sets'}
#Create the Bar Plot
p_2 <- ggplot(final_df_down, 
       aes(x = Gene_set, y = log10p.val)) + 
  theme_bw() + 
  geom_bar(stat = "identity", fill = "steelblue") +
  coord_flip() 

p_2 + ggtitle ("Signficantly downregulated Gene sets (Normal vs Tumor") +
  ylab("-log10pvalue")
```

```{r, warning=FALSE}
#Save as PDF
ggsave(
  file = "Downregulated_genes_bar.pdf",
  plot = last_plot(),
  path = NULL,
  scale = 1,
  width = 15,
  height = 10,
  units = "in",
  dpi = "retina",
  limitsize = TRUE,
  bg = NULL,
)
```


# Notes

GSEA
1. Detect general trend without setting limit on differential genes
2. Kolmogorov-Smirnov-Test, multiple hypotheses testing correction 
3. MSigDB helps finding other related studies

GAGE (Generally Applicable Gene-set Enrichment) [@GAGE]
1. applies to datasets with any number of sample and is based on a parametric gene randomization procedure [@GAGE]
2. Uses log-based fold changes as per gene statistics
3. assumes gene set comes from a different distribution than the background and unses two-sample t-test to account for the gene set specificic variance as well as the background variance
4. adjusts for different experimental designs and sample sizes by decomposing roup-on group comparison into one-on-one comparisons between samples from different groups
- derives a global p-value using a meta-test on the p-values from these comparisons for each gene set
5. separates experimentally perturbed gene sets (from literature, single directions) and canonical pathways (from pathway databases, both directions)


# Useful links:
1. Gage
<http://bioconductor.org/packages/release/bioc/html/gage.html>
2. msigdbr
<https://cran.r-project.org/web/packages/msigdbr/vignettes/msigdbr-intro.html>
3. single cell RNA seq GSEA
<https://www.nature.com/articles/s41467-020-15298-6>


#Data set description

Data retrieved from the [TCBA database](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga) and includes log2-normalized read counts of RNA-sequencing samples from Glioblastoma multiforme (GBM) patients or healthy control tissues. 

Tumor tissues were classified as "Primary" or "Recurrent". Control tissues were classified as being "Normal.".

I used the updated curated gene set collection Hallmark (H) from the [MSigDB databse](https://www.gsea-msigdb.org/gsea/msigdb/) (@Libzerzon et al.2015).

# Citations

```{r}
citation("gage")
citation("org.Hs.eg.db")
citation("msigdbr")
```

# Session info

```{r}
sessionInfo()
```
